src/include/api/flat_l2_index.h

```diff
@@ -0,0 +1,296 @@
+  }
+
+  /**
+   * Non-type parameterized base class (for type erasure).
+   */
+  struct index_base {
+    virtual ~index_base() = default;
+
+    [[nodiscard]] virtual std::tuple<FeatureVectorArray, FeatureVectorArray>
+    query(const QueryVectorArray& vectors, size_t top_k) const = 0;
+
+    virtual void update(
+        const FeatureVectorArray&,
+        const std::optional<IdVector>& ids,
+        const std::optional<UpdateOptions>& options) const = 0;
+
+    virtual void update(
+        const URI& vectors_uri,
+        const std::optional<IdVector>& ids,
+        const std::optional<UpdateOptions>& options) const = 0;
+
+    virtual void remove(const IdVector& ids) const = 0;
+
+    virtual size_t dimension() const = 0;
+
+    virtual size_t ntotal() const = 0;
+
+    virtual size_t num_vectors() const = 0;
+  };
+
+  /**
+   * @brief Type-parameterize implementation class.
+   * @tparam T Type of the concrete class that is being type-erased.
+   */
+  template <typename T>
+  struct index_impl : index_base {
+    explicit index_impl(T&& t)
+        : impl_index_(std::forward<T>(t)) {
+    }
+
+    index_impl(
+        const tiledb::Context& ctx,
+        const URI& index_uri,
+        const std::optional<StringMap>& config = std::nullopt)
+        : impl_index_(ctx, index_uri) {
+    }
+
+    template <feature_vector_array V>
+    index_impl(
+        const URI& index_uri,
+        const V& vectors,
+        const IndexOptions& options,
+        const std::optional<StringMap>& config = std::nullopt)
+        : impl_index_(index_uri, vectors, options, config) {
+    }
+
+    // Create from input URI
+    index_impl(
+        const URI& index_uri,
+        const URI& vectors_uri,
+        const IndexOptions& options,
+        std::optional<StringMap> config = std::nullopt)
+        : impl_index_(index_uri, vectors_uri, options, config) {
+    }
+
+    [[nodiscard]] auto query(
+        tiledb::Context ctx, const URI& uri, size_t top_k) const {
+      return impl_index_.query(ctx, uri, top_k);
+    }
+
+    /**
+     * @brief Query the index with the given vectors.  The concrete query
+     * function returns a tuple of arrays, which are type erased and returned as
+     * a tuple of FeatureVectorArrays.
+     * @param vectors
+     * @param k_nn
+     * @return
+     *
+     * @todo Make sure the extents of the returned arrays are used correctly.
+     */
+    [[nodiscard]] std::tuple<FeatureVectorArray, FeatureVectorArray> query(
+        const QueryVectorArray& vectors, size_t k_nn) const override {
+      // @todo using index_type = size_t;
+
+      auto dtype = vectors.feature_type();
+
+      // @note We need to maintain same layout -> or swap extents
+      switch (dtype) {
+        case TILEDB_FLOAT32: {
+          auto qspan = MatrixView<float, stdx::layout_left>{
+              (float*)vectors.data(),
+              extents(vectors)[0],
+              extents(vectors)[1]};  // @todo ??
```

Comment:
There are quite a few of these `// @todo ??` comments - would suggest to add comments about what the todo is saying, i.e. `// @todo Update to __`, as it's not clear just from reading the code.